### PR TITLE
[SPG-561]: Adding variable resource

### PR DESCRIPTION
**Pull request by Y0709**

### CODE REVIEW

File: internal/service/platform/variables/variables.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	resource := &schema.Resource{
+		Description: "Resource for creating a Harness Variables.",
+
+		ReadContext:   resourceVariablesRead,
+		UpdateContext: resourceVariablesCreateOrUpdate,
+		DeleteContext: resourceVariablesDelete,
+		CreateContext: resourceVariablesCreateOrUpdate,
+		Importer:      helpers.MultiLevelResourceImporter,
+
+		Schema: map[string]*schema.Schema{
```
Comment (RajBaviskar) [inline internal/service/platform/variables/variables.go:24]: add identifier, as required

**Review by RajBaviskar — APPROVED** [review]

File: internal/service/platform/variables/data_source_variables.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	helpers.SetMultiLevelDatasourceSchema(resource.Schema)
```
Comment (rathodmeetsatish) [inline internal/service/platform/variables/data_source_variables.go:54]: please check SetMultiLevelDatasourceSchema identifier field is defined here as well

File: internal/service/platform/variables/variables.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+							Type:        schema.TypeString,
+							Required:    true,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	helpers.SetMultiLevelResourceSchema(resource.Schema)
```
Comment (rathodmeetsatish) [inline internal/service/platform/variables/variables.go:58]: identifier is defined here as well please check SetMultiLevelResourceSchema

File: internal/service/platform/variables/variables.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+		Schema: map[string]*schema.Schema{
+			"identifier": {
+				Description: "Unique identifier of the resource",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"type": {
+				Description: "Type of Variable",
+				Type:        schema.TypeString,
```
Comment (rathodmeetsatish) [inline internal/service/platform/variables/variables.go:32]: if it is enum please add validatefunc field for ref check 
ValidateFunc: validation.StringInSlice([]string{"Reference", "Inline"}, false),
in secret text resource

File: internal/service/platform/variables/variables_test.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+	}
+
+	if resp.Data == nil {
+		return nil, nil
+	}
+
+	return resp.Data.Variable, nil
+}
+
+func testAccGetResourceVariablesOrgLevel(resourceName string, state *terraform.State) (*nextgen.VariableDto, error) {
```
Comment (rathodmeetsatish) [inline internal/service/platform/variables/variables_test.go:110]: why do we need separate method to get org level variable
for ref check testAccGetRoleAssignments method

**Comment by rathodmeetsatish** [comment]
please add a .txt file corresponding to your pr number and define the changes done in this pr please refer 
https://github.com/harness/terraform-provider-harness/pull/250/files#diff-3dcea50f1bab104aad6917af874e70be5750a96acf4be00487e48bd08ce74245

File: internal/provider/provider.go [hunk 6]
```diff
@@ -145,6 +146,7 @@ func Provider(version string) func() *schema.Provider {
 				"harness_platform_service_account":               service_account.DataSourceServiceAccount(),
 				"harness_platform_triggers":                      triggers.DataSourceTriggers(),
 				"harness_platform_role_assignments":              role_assignments.DataSourceRoleAssignments(),
+        "harness_platform_variables":                     variables.DataSourceVariables(),
```
Comment (rijajoo) [inline internal/provider/provider.go:149]: @Y0709 please format accordingly.

File: internal/provider/provider.go [hunk 7]
```diff
@@ -208,6 +210,7 @@ func Provider(version string) func() *schema.Provider {
 				"harness_platform_service_account":               service_account.ResourceServiceAccount(),
 				"harness_platform_triggers":                      triggers.ResourceTriggers(),
 				"harness_platform_role_assignments":              role_assignments.ResourceRoleAssignments(),
+        "harness_platform_variables":                     variables.ResourceVariables(),
```
Comment (rijajoo) [inline internal/provider/provider.go:213]: Same here.

File: internal/provider/provider.go (hunk 6 above)
Comment (Y0709) [reply internal/provider/provider.go]: done

File: internal/provider/provider.go (hunk 7 above)
Comment (Y0709) [reply internal/provider/provider.go]: done

**Comment by rijajoo** [comment]
Terraform build is failing. Please check

File: .changelog/251.txt [hunk 8]
```diff
@@ -0,0 +1,3 @@
+```release-note:new-resource
+variables - Added variable resource in Harness terraform provider
```
Comment (rathodmeetsatish) [inline .changelog/251.txt:2]: platform_variables

Comment (Y0709) [reply .changelog/251.txt]: done

**Pull request merged**